ddtrace/tracer/post_processor.go

```diff
@@ -0,0 +1,111 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022 Datadog, Inc.
+
+package tracer
+
+import (
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+var _ ddtrace.ReadWriteSpan = (*readWriteSpan)(nil)
+
+// readWriteSpan wraps span and implements the ddtrace.ReadWriteSpan interface.
+type readWriteSpan struct {
+	*span
+}
+
+// Tag returns the tag value held by the given key.
+func (s readWriteSpan) Tag(key string) interface{} {
+	s.Lock()
+	defer s.Unlock()
+
+	switch key {
+	// String.
+	case ext.SpanName:
+		return s.Name
+	case ext.ServiceName:
+		return s.Service
+	case ext.ResourceName:
+		return s.Resource
+	case ext.SpanType:
+		return s.Type
+	// Bool.
+	case ext.AnalyticsEvent:
+		return s.Metrics[ext.EventSampleRate] == 1.0
+	case ext.ManualDrop:
+		return s.Metrics[keySamplingPriority] == -1
+	case ext.ManualKeep:
+		return s.Metrics[keySamplingPriority] == 2
+	// Metrics.
+	case ext.SamplingPriority, keySamplingPriority:
+		if val, ok := s.Metrics[keySamplingPriority]; ok {
+			return val
+		}
+		return nil
+	}
+	if val, ok := s.Meta[key]; ok {
+		return val
+	}
+	if val, ok := s.Metrics[key]; ok {
+		return val
+	}
+	return nil
+}
+
+// IsError reports wether s is an error.
+func (s readWriteSpan) IsError() bool {
+	s.Lock()
+	defer s.Unlock()
+
+	return s.Error == 1
+}
+
+// SetOperationName is a no-op.
+func (s readWriteSpan) SetOperationName(operationName string) {
+	s.SetTag(ext.SpanName, operationName)
+}
+
+// SetTag adds a set of key/value metadata to the span. Setting metric aggregator tags
+// (name, env, service, version, resource, http.status_code and keyMeasured) or modifying
+// the sampling priority in the processor is not allowed.
+func (s readWriteSpan) SetTag(key string, value interface{}) {
```

Comment:
Can we not re-use code from places like this? https://github.com/DataDog/dd-trace-go/blob/50bd7cebbd482284c99e41321495ac610990269a/ddtrace/tracer/span.go#L106
(instead of duplicating it)

Reply:
We do 🙂. There is a new method `setTagLocked` ([link](https://github.com/DataDog/dd-trace-go/blob/a96015d3909a2c8f319f63ec701c92c548f15e96/ddtrace/tracer/span.go#L118)) which contains the code that can be re-used between
`(s readWriteSpan) SetTag` and `(s *span) SetTag`. Now both these methods call `setTagLocked` ([span](https://github.com/DataDog/dd-trace-go/blob/a96015d3909a2c8f319f63ec701c92c548f15e96/ddtrace/tracer/span.go#L115) and [readWriteSpan](https://github.com/DataDog/dd-trace-go/blob/a96015d3909a2c8f319f63ec701c92c548f15e96/ddtrace/tracer/post_processor.go#L90)). 

The reasons for having `(s readWriteSpan) SetTag` method that shadows `(s *span) SetTag` are:
- In the processor, changing some fields is not allowed, so `(s readWriteSpan) SetTag` will be a no-op if the user tries to change these fields.
- `(s *span) SetTag` is a no-op on finished spans, but `(s readWriteSpan) SetTag` is not.